kotlinx-coroutines-core/jvm/src/CoroutineContext.kt

```diff
@@ -10,35 +10,68 @@ import kotlin.coroutines.jvm.internal.CoroutineStackFrame
 
 /**
  * Creates context for the new coroutine. It installs [Dispatchers.Default] when no other dispatcher nor
- * [ContinuationInterceptor] is specified, and adds optional support for debugging facilities (when turned on).
+ * [ContinuationInterceptor] is specified, and adds optional support for
+ * copyable thread context [elements][CopyableThreadContextElement] and  debugging facilities (when turned on).
  *
  * See [DEBUG_PROPERTY_NAME] for description of debugging facilities on JVM.
+ *
  */
 @ExperimentalCoroutinesApi
 public actual fun CoroutineScope.newCoroutineContext(context: CoroutineContext): CoroutineContext {
-    val combined = coroutineContext.foldCopiesForChildCoroutine() + context
+    val combined = foldCopies(context)
     val debug = if (DEBUG) combined + CoroutineId(COROUTINE_ID.incrementAndGet()) else combined
     return if (combined !== Dispatchers.Default && combined[ContinuationInterceptor] == null)
         debug + Dispatchers.Default else debug
 }
 
-/**
- * Returns the [CoroutineContext] for a child coroutine to inherit.
- *
- * If any [CopyableThreadContextElement] is in the [this], calls
- * [CopyableThreadContextElement.copyForChildCoroutine] on each, returning a new [CoroutineContext]
- * by folding the returned copied elements into [this].
+private val hasCopyableElements: (Boolean, CoroutineContext.Element) -> Boolean = { result, it ->
+    result || it is CopyableThreadContextElement<*, *>
+}
+
+/*
+ * Folds two contexts if there is need to.
  *
- * Returns [this] if `this` has zero [CopyableThreadContextElement] in it.
+ * The algorithm is the following:
+ * *
+ * *
+ * *
  */
-private fun CoroutineContext.foldCopiesForChildCoroutine(): CoroutineContext {
-    val hasToCopy = fold(false) { result, it ->
-        result || it is CopyableThreadContextElement<*>
+private fun CoroutineScope.foldCopies(appendContext: CoroutineContext): CoroutineContext {
+    // Do we have something to copy left-hand side?
+    val hasElementsLeft = coroutineContext.fold(false, hasCopyableElements)
+    val hasElementsRight = appendContext.fold(false, hasCopyableElements)
```

Comment:
It is not entirely clear why do you `hasElementsRight` here. It looks that even if the right has copyable element (but the left has not), then effect of this code is a simple `coroutineContext + appendContext`, so you can still take the fast path.

Reply:
I'm still exploring whether we should always do copy when there is an element on the right-hand side.